### ISSUE TITLE
Summary of 'halts' in standard modules

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#9386]
Many errors (or unimplemented features) in the standard modules just halt today. Note that "halt" includes:
 - calls to `halt()`
 - calls to `assert()`
 - uses of `__primitive("chpl_error")`
 - uses of `try!`

Eventually, we want to prevent our standard modules from halting, and either implement better code that doesn't need to halt, or propagate errors up to users via error handling. Below is a summary of halting code in standard modules. The biggest/hardest cases to handle with be:
 - I/O (which should probably be split into a separate issue)
 - DateTime (just has a lot of things that halt)
 - Math (messy and lots of extern calls, not obvious how we want to handle errors in general here)

Some general open questions:
 - What should we do for "abstract base methods"/"pure virtual functions"?
 - What's our general thinking/strategy about out-errors?

---
Summary of "halts" in standard modules:

- [ ] Barriers.chpl
  - halts in base barrier for "pure virtual function"
  - halts for too many calls to a barrier/notify
  - halts for reset on sync barrier (or trying to make sync reusable)

* [ ] BigInteger.chpl
  * halt and out error init (from string) if bad format
  * halt in cast for negative bigint to uint

- [ ] DateTime.chpl
  - halt for gettimeofday failures
  - halt for month not between 1 and 12
  - halt for day not between 1 and daysInMonth
  - halt for invalid year/month/day range
  - halt for ordinal out of range
  - halt for mismatched owned/borrowed/unmanaged wrapping
  - halt for "unreachable"
  - halt for utcoffset with "naive" datetime?
  - halt for dst() called with nil tzinfo
  - halt for days overflow
  - halt for abstract base method (pure virtual func)
  - couple of try! for I/O and string casting
  - 3 misc asserts

* [ ] DynamicIters.chpl
  * halt for empty range
  * halt for negative numTasks
  * lots of asserts for parDim/rank and chunksize

- [ ] FileSystem.chpl
  - lots of "out error" overloads
  - halt for "glob zipped with something too big"
  - halt for "cannot move dir into itself"
  - __primitive("chpl_error") for "unhandled error in glob"

* [ ] IO.chpl (probably split into separate issue)
  * tons of "out error" overloads
  * lots of halts
  * lots of try!
  * related to https://github.com/chapel-lang/chapel/issues/8694 and https://github.com/chapel-lang/chapel/issues/7261?

- [ ] List.chpl
  - halt for pop front on empty list

* [ ] Math.chpl (need to figure out a general strategy for this one)
  * Weird/old/outdated "Error handling" comment at the top
  * halt for log of non-positive num
  * halt for bessel functions on non-positive num
  * this module is a thin shim over calls to c math routines
    * probably need to figure out what we want to do for "errors" here

- [ ] Path.chpl
  - uses "out error" overloads

* [ ] Random.chpl
  * halt for non-positive skipToNth
  * halt for even seed
  * halt for seed not been 0 and 2**46

- [ ] Regexp.chpl
  - assert for "nmatches >= captures.size"
  - uses "out error" overload

* [ ] Spawn.chpl
  * lots of "out errors"

- [ ] Time.chpl
  - halt for start called on a non-stopped timer
  - halt called for stop called on a non-started timer
  - halt for "internal error in module Time"
  - try! for sleep() with negative value 

* [ ] Types.chpl
  * halt for safeCasts

- [ ] UtilReplicatedVar.chpl
  - asserts for some domain checks

---
No obvious halts/asserts/whatever:

- Assert.chpl (this just provides the assert procs)
- BitOps.chpl
- CommDiagnostics.chpl
- Error.chpl (just a "deprecated" forward to SysError)
- GMP.chpl
- Help.chpl
- Memory.chpl
- Reflection.chpl
- Sys.chpl
- SysBasic.chpl
- SysError.chpl

**Comment by ronawho** [comment]
All uses of `__primitive("chpl_error")` and most uses of `halt()` have been removed or replaced with HaltWrappers in the standard modules. The only cases that remain are:
 - DateTime -- blocked on https://github.com/chapel-lang/chapel/issues/9941
 - log2() -- blocked on https://github.com/chapel-lang/chapel/issues/10292

There are some intentional/acceptable uses of `try!` in DateTime, but I have not looked at all the cases in I/O

We have also decided that asserts are acceptable
